File: 23-meeting-rooms/solution.cpp
// Meeting Rooms II - Minimum Number of Meeting Rooms
// Time: O(n log n), Space: O(n)

#include <iostream>
#include <vector>
#include <queue>
#include <algorithm>

using namespace std;

class Solution {
public:
    int minMeetingRooms(vector<vector<int>>& intervals) {
        if (intervals.empty()) return 0;
        
        sort(intervals.begin(), intervals.end());
        priority_queue<int, vector<int>, greater<int>> heap;
        heap.push(intervals[0][1]);
        
        for (int i = 1; i < intervals.size(); i++) {
            if (intervals[i][0] >= heap.top()) {
                heap.pop();
            }
            heap.push(intervals[i][1]);
        }
        
        return heap.size();
    }
};

// Test
int main() {
    Solution sol;
    vector<vector<int>> intervals1 = {{0, 30}, {5, 10}, {15, 20}};
    cout << sol.minMeetingRooms(intervals1) << endl;  // 2
    
    vector<vector<int>> intervals2 = {{7, 10}, {2, 4}};
    cout << sol.minMeetingRooms(intervals2) << endl;  // 1
    
    return 0;
}

File: 23-meeting-rooms/solution.go
// Meeting Rooms II - Minimum Number of Meeting Rooms
// Time: O(n log n), Space: O(n)

package main

import (
	"container/heap"
	"fmt"
	"sort"
)

type IntHeap []int

func (h IntHeap) Len() int           { return len(h) }
func (h IntHeap) Less(i, j int) bool { return h[i] < h[j] }
func (h IntHeap) Swap(i, j int)      { h[i], h[j] = h[j], h[i] }

func (h *IntHeap) Push(x interface{}) {
	*h = append(*h, x.(int))
}

func (h *IntHeap) Pop() interface{} {
	old := *h
	n := len(old)
	x := old[n-1]
	*h = old[0 : n-1]
	return x
}

type Solution struct{}

func (s *Solution) minMeetingRooms(intervals [][]int) int {
	if len(intervals) == 0 {
		return 0
	}
	
	sort.Slice(intervals, func(i, j int) bool {
		return intervals[i][0] < intervals[j][0]
	})
	
	h := &IntHeap{intervals[0][1]}
	heap.Init(h)
	
	for i := 1; i < len(intervals); i++ {
		if intervals[i][0] >= (*h)[0] {
			heap.Pop(h)
		}
		heap.Push(h, intervals[i][1])
	}
	
	return h.Len()
}

// Test
func main() {
	sol := &Solution{}
	fmt.Println(sol.minMeetingRooms([][]int{{0, 30}, {5, 10}, {15, 20}})) // 2
	fmt.Println(sol.minMeetingRooms([][]int{{7, 10}, {2, 4}}))            // 1
}
